oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,438 @@
+
+    if (!SUPPORTED_FORMAT_VERSION.equals(formatVersion)) {
+      throw new IOException(
+          String.format("Only format version %s is supported.", SUPPORTED_FORMAT_VERSION));
+    }
+
+    URI tokenServerUriFromCreds = null;
+    try {
+      tokenServerUriFromCreds = new URI(tokenServerUriStringFromCreds);
+    } catch (URISyntaxException e) {
+      throw new IOException("Token server URI specified in 'token_uri' could not be parsed.");
+    }
+
+    GdchCredentials.Builder builder =
+        GdchCredentials.newBuilder()
+            .setProjectId(projectId)
+            .setPrivateKeyId(privateKeyId)
+            .setTokenServerUri(tokenServerUriFromCreds)
+            .setServiceIdentityName(serviceIdentityName)
+            .setHttpTransportFactory(transportFactory);
+
+    return fromPkcs8(privateKeyPkcs8, builder);
+  }
+
+  /**
+   * Internal constructor.
+   *
+   * @param privateKeyPkcs8 RSA private key object for the service account in PKCS#8 format.
+   * @param builder A builder for GdchCredentials.
+   * @return an instance of GdchCredentials.
+   */
+  static GdchCredentials fromPkcs8(String privateKeyPkcs8, GdchCredentials.Builder builder)
+      throws IOException {
+    PrivateKey privateKey = OAuth2Utils.privateKeyFromPkcs8(privateKeyPkcs8);
+    builder.setPrivateKey(privateKey);
+
+    return new GdchCredentials(builder);
+  }
+
+  /**
+   * Create a copy of GDCH credentials with the specified audience.
+   *
+   * @param apiAudience The intended audience for GDCH credentials.
+   */
+  public GdchCredentials createWithGdchAudience(URI apiAudience) throws IOException {
+    Preconditions.checkNotNull(
+        apiAudience, "Audience are not configured for GDCH service account credentials.");
+    return this.toBuilder().setGdchAudience(apiAudience).build();
+  }
+
+  /**
+   * Refresh the OAuth2 access token by getting a new access token using a JSON Web Token (JWT).
+   *
+   * <p>For GDCH credentials, this class creates a self-signed JWT, and sends to the GDCH
+   * authentication endpoint (tokenServerUri) to exchange an access token for the intended api
+   * audience (apiAudience).
+   */
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    Preconditions.checkNotNull(
+        this.apiAudience,
+        "Audience are not configured for GDCH service account. Specify the "
+            + "audience by calling createWithGDCHAudience.");
+
+    JsonFactory jsonFactory = OAuth2Utils.JSON_FACTORY;
+    long currentTime = clock.currentTimeMillis();
+    String assertion = createAssertion(jsonFactory, currentTime, getApiAudience());
+
+    GenericData tokenRequest = new GenericData();
+    tokenRequest.set("grant_type", OAuth2Utils.TOKEN_TYPE_TOKEN_EXCHANGE);
+    tokenRequest.set("assertion", assertion);
+    UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
+
+    HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
```

Comment:
what about ca cert?

Reply:
Added a class implementing HttpTransport which reads ca cert and creates the transport with trust store.